pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -352,15 +357,17 @@ protected void readMoreEntries(Consumer consumer) {
             if (log.isDebugEnabled()) {
                 log.debug("[{}-{}] Schedule read of {} messages", name, consumer, messagesToRead);
             }
-            havePendingRead = true;
-            if (consumer.readCompacted()) {
-                topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
-                        this, consumer);
-            } else {
-                ReadEntriesCtx readEntriesCtx =
-                        ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
-                cursor.asyncReadEntriesOrWait(messagesToRead,
-                        bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
+            synchronized (this) {
+                havePendingRead = true;
+                if (consumer.readCompacted()) {
+                    topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
+                            this, consumer);
+                } else {
+                    ReadEntriesCtx readEntriesCtx =
+                            ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
+                    cursor.asyncReadEntriesOrWait(messagesToRead,
+                            bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
+                }
```

Comment:
Could you explain why did you add the `synchronized` block here? It's redundant because the only modification on this object is `havePendingRead` but it's volatile so the write operation is thread safe. While the following logic doesn't rely on the value of `havePendingRead` and doesn't modify any state of `PersistentDispatcherSingleActiveConsumer` itself.

Reply:
/cc @Jason918 @codelipenghui @nodece @congbobo184 WDYT?

Reply:
oh, you are right. I was originally concerned that calling the cursor method is not thread-safe, I looked at the code again, the processing here is originally the way of optimistic locking. Thank your review.

@congbobo184 can you help make sure?

Reply:
because the readMoreEntries isn't a synchronized method. If any other thread invoke `readMoreEntries` and other thread 
https://github.com/shibd/pulsar/blob/83a983dd21b48a06ef23db8a07b5536d60de8ca8/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L287-L313

it may use the new epoch and use the old readPostion, it will lose message .so add a `synchronized` block

Reply:
But why should we synchronize the following code block, not the whole `readMoreEntries` or just `calculateToRead` method?

```java
            havePendingRead = true;
            if (consumer.readCompacted()) {
                topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
                        this, consumer);
            } else {
                ReadEntriesCtx readEntriesCtx =
                        ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
                cursor.asyncReadEntriesOrWait(messagesToRead,
                        bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
            }
```

Reply:
> it may use the new epoch and use the old readPostion

I don't get what the `old readPosition` represents.

I mean, why not

```java
    protected synchronized void readMoreEntries(Consumer consumer) {
        /* ... */
    }
```

or

```java
    protected void readMoreEntries(Consumer consumer) {
        // consumer can be null when all consumers are disconnected from broker.
        // so skip reading more entries if currently there is no active consumer.
        if (null == consumer) {
            return;
        }

        synchronized (this) {
            if (consumer.getAvailablePermits() > 0) {
                Pair<Integer, Long> calculateResult = calculateToRead(consumer);
                /* ... */
            }
        }
```

or synchronize other code block, but just the code block I pointed out before:

```java
            havePendingRead = true;
            if (consumer.readCompacted()) {
                topic.getCompactedTopic().asyncReadEntriesOrWait(cursor, messagesToRead, isFirstRead,
                        this, consumer);
            } else {
                ReadEntriesCtx readEntriesCtx =
                        ReadEntriesCtx.create(consumer, consumer.getConsumerEpoch());
                cursor.asyncReadEntriesOrWait(messagesToRead,
                        bytesToRead, this, readEntriesCtx, topic.getMaxReadPosition());
            }
```

**Please point out which variables need synchronizing**. As far as I see:
- `messagesToRead` and `bytesToRead` are computed by `calculateToRead`, which is not protected by a synchronized block.
- `consumer`, `readEntriesCtx` are related to the `consumer` argument of `readMoreEntries`, and I think these code doesn't change any state of the consumer.
- `CompactTopicImpl#asyncReadEntriesOrWait` is a synchronized method.
-  `ManagedCursorImpl#asyncReadEntriesOrWait` is not a synchronized method but it's thread safe, otherwise there is no need to use an atomic reference (`WAITING_READ_OP_UPDATER`) in it.

Reply:
The first this PR only fix the new consumer epoch in `OpReadEntry` and OpReadEntry use old readPosition to read `ManagedLedger`.

steps to reproduce:
1. one thread invoke this method
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L296
and then change the cosnumerEpoch from 0 -> 1
this method hasn't rewind the cursor 
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L319
2. one thread invoke readMoreEntry
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L361-L364
now the cursor readPostion hasn't been changed but consumerEpoch has changed. this readOp read the messages, the client will not filter. If you want to redeliver 1-10, but send you the messages is from 10-15 and cumulative ack then client broken down 1-10 messages will loss.

The second is because https://github.com/apache/pulsar/pull/15046 changed the code. 

```
        if (!havePendingRead) {
            cursor.rewind();
            if (log.isDebugEnabled()) {
                log.debug("[{}-{}] Cursor rewinded, redelivering unacknowledged messages. ", name, consumer);
            }
            readMoreEntries(consumer);
        } else {
            log.info("[{}-{}] Ignoring reDeliverUnAcknowledgedMessages: cancelPendingRequest on cursor failed", name,
                    consumer);
        }
```
to

> https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L317-L323

so`havePendingRead` also doesn't need to add the synchronized block. any other code also doesn't need to move into the synchronized block.

```
ManagedCursorImpl#asyncReadEntriesOrWait is not a synchronized method but it's thread safe, otherwise there is no need to use an atomic reference (WAITING_READ_OP_UPDATER) in it.
```
asyncReadEntriesOrWait is thread-safe but it can't solve 
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java#L48 use old readPostion

and 
https://github.com/apache/pulsar/blob/48090f36e3c37630d39f67504cf77656bd386b9b/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java#L617 use new consumer epoch.

The third, because this is the critical path of reading, we try to reduce the code in the lock as much as possible, other code doesn't need to add to the synchronized block

Reply:
@congbobo184 can we just make `consumer.getConsumerEpoch()`and `consumer.setConsumerEpoch()` keep synchronized?

Reply:
@codelipenghui  I think this lock is mainly intended for `cursor.rewind()` in `internalRedeliverUnacknowledgedMessages`, which makes the epoch change and cursor rewind atomic.